processing/src/main/java/io/druid/segment/filter/Filters.java

```diff
@@ -208,73 +214,167 @@ public static ImmutableBitmap matchPredicate(
-            selector.getBitmapFactory().makeEmptyImmutableBitmap(),
-            selector.getNumRows()
-        );
-      } else {
-        return selector.getBitmapFactory().makeEmptyImmutableBitmap();
-      }
+      return predicate.apply(null) ? allTrue(selector) : allFalse(selector);
     }
 
     // Apply predicate to all dimension values and union the matching bitmaps
     final BitmapIndex bitmapIndex = selector.getBitmapIndex(dimension);
-    return selector.getBitmapFactory().union(
-        new Iterable<ImmutableBitmap>()
+    return selector.getBitmapFactory()
+                   .union(makePredicateQualifyingBitmapIterable(bitmapIndex, predicate, dimValues));
+  }
+
+  /**
+   * Return an estimated selectivity for bitmaps of all values matching the given predicate.
+   *
+   * @param columnSelector column selector
+   * @param dimension      dimension to look at
+   * @param indexSelector  bitmap selector
+   * @param predicate      predicate to use
+   *
+   * @return estimated selectivity
+   *
+   * @see Filters#matchPredicate(String, BitmapIndexSelector, Predicate)
+   */
+  static double estimatePredicateSelectivity(
+      final ColumnSelector columnSelector,
+      final String dimension,
+      final BitmapIndexSelector indexSelector,
+      final Predicate<String> predicate
+  )
+  {
+    Preconditions.checkNotNull(dimension, "dimension");
+    Preconditions.checkNotNull(indexSelector, "selector");
+    Preconditions.checkNotNull(predicate, "predicate");
+
+    // Missing dimension -> match all rows if the predicate matches null; match no rows otherwise
+    final Indexed<String> dimValues = indexSelector.getDimensionValues(dimension);
+    if (dimValues == null || dimValues.size() == 0) {
+      return predicate.apply(null) ? 1. : 0.;
+    }
+
+    // Apply predicate to all dimension values and union the matching bitmaps
+    final BitmapIndex bitmapIndex = indexSelector.getBitmapIndex(dimension);
+    return estimatePredicateSelectivity(
+        columnSelector,
+        dimension,
+        makePredicateQualifyingBitmapIterable(
+            bitmapIndex,
+            predicate,
+            dimValues
+        ),
+        indexSelector.getNumRows()
+    );
+  }
+
+  static double estimatePredicateSelectivity(
+      ColumnSelector columnSelector,
+      String dimension,
+      Iterable<ImmutableBitmap> bitmaps,
+      long totalNumRows
+  )
+  {
+    final ColumnCapabilities columnCapabilities = columnSelector.getColumn(dimension).getCapabilities();
+    return estimatePredicateSelectivity(
+        bitmaps,
+        totalNumRows,
+        columnCapabilities == null || columnCapabilities.hasMultipleValues()
+    );
+  }
+
+  static double estimatePredicateSelectivity(
+      Iterable<ImmutableBitmap> bitmaps,
+      long totalNumRows,
+      boolean isMultiValueDimension
+  )
+  {
+    long numMatchedRows = 0;
+    final List<ImmutableBitmap> bitmapList = Lists.newArrayList(bitmaps);
```

Comment:
We try to avoid materializing all bitmaps in memory at the same time. See https://github.com/druid-io/druid/pull/3642#discussion_r86036326. Maybe you need to pass around `IntIterator` + a lazy int -> `ImmutableBitmap` function, and use it in `computeNonOverlapRatioFromBitmapSamples()`.

cc @gianm